[OpenMP] Using `SimpleVLA` to handle vla usage in ompt-general.cpp.

DanielCChen opened the pull request:
The `openmp` runtime failed to build on LoP with LLVM18 on LoP due to the addition of `-Wvla-cxx-extension` as
```
llvm-project/openmp/runtime/src/ompt-general.cpp:711:15: error: variable length arrays in C++ are a Clang extension [-Werror,-Wvla-cxx-extension]
  711 |   int tmp_ids[ids_size];
      |               ^~~~~~~~
llvm-project/openmp/runtime/src/ompt-general.cpp:711:15: note: function parameter 'ids_size' with unknown value cannot be used in a constant expression
llvm-project/openmp/runtime/src/ompt-general.cpp:704:65: note: declared here
  704 | OMPT_API_ROUTINE int ompt_get_place_proc_ids(int place_num, int ids_size,
      |                                                                 ^
1 error generated.
```

This patch is to ignore the checking against this usage.

DanielCChen committed:
[OpenMP] Ignore the error introduced by -Wvla-cxx-extension on ompt-geenral.cpp.

jhuber6 commented:
I would prefer we remove the VLA access, but I don't know how difficult it is for this case.

shiltian reviewed (changes requested):
No, this is wrong. I previous added a VLA-equal template thing in `libomp` and replaced all uses there. Can you move it somewhere common such that ompt can also use it?

DanielCChen commented:
> No, this is wrong. I previous added a VLA-equal template thing in `libomp` and replaced all uses there. Can you move it somewhere common such that ompt can also use it?

Thanks for the pointer! @shiltian 
I am pretty sure this is the change you are referring to (https://github.com/llvm/llvm-project/commit/eaab947a8aa39002e8bdaa82be08cbc31e116a11).

I replaced the offending code in `ompt-general.cpp` as
```
//  int tmp_ids[ids_size];
  SimpleVLA<int> tmp_ids(ids_size);
```
and it fixed the issue. 

Since `kmp_utils.h` only has `SimpleVLA`,  I can either just include it in `ompt-general.cpp`, or rename it to something like `omp_utility.h` to be more general. Thoughts?

DanielCChen committed:
[OpenMP] to address the review comments to use template SimpleVLA.

shiltian reviewed (approved):
I thought ompt source file was in a separate directory. :-)

shiltian commented on openmp/runtime/src/ompt-general.cpp:
```diff
@@ -23,6 +23,7 @@
 #if KMP_OS_UNIX
 #include <dlfcn.h>
 #endif
+#include "kmp_utils.h"
```
put it in the right order?

DanielCChen replied:
> put it in the right order?

Could you please be more specific? @shiltian

hubert-reinterpretcast replied:
@DanielCChen, I think https://llvm.org/docs/CodingStandards.html#include-style says that this header should be moved up. Indeed, it wouldn't fit under the "system include files" comment.

I don't know what to say about the include of `ompt-specific.cpp` (yes, `.cpp`!) below.

shiltian replied:
Yes. That is what I meant.

> I don't know what to say about the include of ompt-specific.cpp (yes, .cpp!) below.

Me neither. ;-) That is before `libomp` was contributed to LLVM.

DanielCChen replied:
Thanks. I will move it before the system #include then.

DanielCChen committed:
Move LLVM headers before system headers.
